Convert deprecation policy and maintainer guide to markdown (#11218)
Qiskit/qiskit: DEPRECATION.md

Beware that users will often be using functions, classes and methods that we,
the Qiskit developers, may consider internal or not widely used.  Do not make
assumptions that "this is buried, so nobody will be using it"; if it is public,
it is subject to the policy.  The only exceptions here are functions and modules
that are explicitly internal, *i.e.* those whose names begin with a leading
underscore (`_`).

The guiding principles are:

- we must not remove or change code without active warnings for least three
  months or two complete version cycles;

- there must always be a way to achieve valid goals that does not issue any
  warnings;

- never assume that a function that isn't explicitly internal isn't in use;

- all deprecations, changes and removals are considered API changes, and can
  only occur in minor releases not patch releases, per the [stable branch policy](https://github.com/Qiskit/qiskit/blob/main/MAINTAINING.md#stable-branch-policy).

## Removing a feature

When removing a feature (for example a class, function or function parameter),
we will follow this procedure:

- The alternative path must be in place for one minor version before any

>    warnings are issued.  For example, if we want to replace the function `foo()`

Reviewer: In all the bullets in this list, the continuation-line indentation looks like it's got an extra space in it.  Not that it affects the display, it just makes it a bit harder to update.

   with `bar()`, we must make at least one release with both functions before
   issuing any warnings within `foo()`.  You may issue

>    `PendingDeprecationWarning`\ s from the old paths immediately.

Reviewer: ```suggestion
   `PendingDeprecationWarning`s from the old paths immediately.
```

the escaped space isn't necessary in GitHub markdown, and I'm not sure regular markdown interprets the escape.

  removing this keyword argument after you have made the behavior change.  This
  will take at least six months to go through both cycles.

- If you need to change the return type of a function, consider adding a new
  function that returns the new type, and then follow the procedures for
  deprecating the old function.

- If you need to accept a new input that you cannot distinguish from an existing
  possibility because of its type, consider letting it be passed by a different
  keyword argument, or add a second function that only accepts the new form.

## Issuing deprecation warnings

The proper way to raise a deprecation warning is to use the decorators `@deprecate_arg` and
`@deprecate_func` from `qiskit.utils.deprecation`. These will generate a standardized message and
and add the deprecation to that function's docstring so that it shows up in the docs.

```python
from qiskit.utils.deprecation import deprecate_arg, deprecate_func

@deprecate_func(since="0.24.0", additional_msg="No replacement is provided.")
def deprecated_func():
    pass

@deprecate_arg("bad_arg", new_alias="new_name", since="0.24.0")
def another_func(bad_arg: str, new_name: str):
    pass

> 

Reviewer: ````suggestion

```python
from qiskit.utils.deprecation import deprecate_arg, deprecate_func

@deprecate_func(since="0.24.0", additional_msg="No replacement is provided.")
def deprecated_func():
    pass

@deprecate_arg("bad_arg", new_alias="new_name", since="0.24.0")
def another_func(bad_arg: str, new_name: str):
    pass
```

````

> ```
>    from qiskit.utils.deprecation import deprecate_arg, deprecate_func
> 
>     @deprecate_func(since="0.24.0", additional_msg="No replacement is provided.")
>     def deprecated_func():
>         pass
> 
>     @deprecate_arg("bad_arg", new_alias="new_name", since="0.24.0")
>     def another_func(bad_arg: str, new_name: str):
>         pass
> ```

Reviewer 2: ```suggestion
```python
from qiskit.utils.deprecation import deprecate_arg, deprecate_func

@deprecate_func(since="0.24.0", additional_msg="No replacement is provided.")
def deprecated_func():
    pass

@deprecate_arg("bad_arg", new_alias="new_name", since="0.24.0")
def another_func(bad_arg: str, new_name: str):
    pass
```
```

> Usually, you should set `additional_msg: str ` with the format `"Instead, use ..."` so that

Reviewer 2: ```suggestion
Usually, you should set `additional_msg: str` with the format `"Instead, use ..."` so that
```

people know how to migrate. Read those functions' docstrings for additional arguments like
`pending: bool` and `predicate`.

If you are deprecating outside the main Qiskit repo, set `package_name` to match your package.
Alternatively, if you prefer to use your own decorator helpers, then have them call
`add_deprecation_to_docstring` from `qiskit.utils.deprecation`.

If `@deprecate_func` and `@deprecate_arg` cannot handle your use case, consider improving
them. Otherwise, you can directly call the `warn` function
from the [warnings module in the Python standard library](https://docs.python.org/3/library/warnings.html),
using the category `DeprecationWarning`.  For example:

> ```

Reviewer 2: ```suggestion
```python
```

import warnings

def deprecated_function():
   warnings.warn(
      "The function qiskit.deprecated_function() is deprecated since "

>       "Qiskit Terra 0.20.0, and will be removed 3 months or more later. "

Reviewer 2: ```suggestion
      "Qiskit 0.44.0, and will be removed 3 months or more later. "
```

      "Instead, you should use qiskit.other_function().",
      category=DeprecationWarning,
      stacklevel=2,
   )
   # ... the rest of the function ...

```

Make sure you include the version of the package that introduced the deprecation
warning (so maintainers can easily see when it is valid to remove it), and what
the alternative path is.

> Take note of the `stackleve`` argument.  This controls which function is

Reviewer 2: ```suggestion
Take note of the `stacklevel` argument.  This controls which function is
```

accused of being deprecated.  Setting `stacklevel=1` (the default) means the
warning will blame the `warn` function itself, while `stacklevel=2` will
correctly blame the containing function.  It is unusual to set this to anything
other than `2`, but can be useful if you use a helper function to issue the
same warning in multiple places.

## Testing deprecated functionality

Whenever you add deprecation warnings, you will need to update tests involving
the functionality.  The test suite should fail otherwise, because of the new
warnings.  We must continue to test deprecated functionality throughout the
deprecation period, to ensure that it still works.

To update the tests, you need to wrap each call of deprecated behavior in its
own assertion block.  For subclasses of `unittest.TestCase` (which all Qiskit
test cases are), this is done by:

> ```

Reviewer 2: ```suggestion
```python
```

   def test_deprecated_function(self):
      with self.assertWarns(DeprecationWarning):
         output = deprecated_function()
      # ... do some things with output ...
      self.assertEqual(output, expected)
```

## Documenting deprecations and breaking changes

It is important to warn the user when your breaking changes are coming.

`@deprecate_arg` and `@deprecate_func` will automatically add the deprecation to the docstring
for the function so that it shows up in docs.

If you are not using those decorators, you should directly add a [Sphinx deprecated directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated):

> ```python
> def deprecated_function():
>     """
>     Short description of the deprecated function.
> 
>     .. deprecated:: 0.44.0
>        The function qiskit.deprecated_function() is deprecated since
>        Qiskit 0.44.0, and will be removed 3 months or more later.
>        Instead, you should use qiskit.other_function().
> 
>     <rest of the docstring>
>     """
>     # ... the rest of the function ...

Reviewer: ````suggestion
```python
def deprecated_function():
    """
    Short description of the deprecated function.

    .. deprecated:: 0.44.0
       The function qiskit.deprecated_function() is deprecated since
       Qiskit 0.44.0, and will be removed 3 months or more later.
       Instead, you should use qiskit.other_function().

    <rest of the docstring>
    """
    # ... the rest of the function ...
```
````

> ```
>    def deprecated_function():
>       """
>       Short description of the deprecated function.
> 
>       .. deprecated:: 0.20.0
>          The function qiskit.deprecated_function() is deprecated since
>          Qiskit Terra 0.20.0, and will be removed 3 months or more later.
>          Instead, you should use qiskit.other_function().
> 
>       <rest of the docstring>
>       """
>       # ... the rest of the function ...
> ```

Reviewer 2: ```suggestion
```python
def deprecated_function():
    """
    Short description of the deprecated function.

    .. deprecated:: 0.44.0
       The function qiskit.deprecated_function() is deprecated since
       Qiskit 0.44.0, and will be removed 3 months or more later.
       Instead, you should use qiskit.other_function().

    <rest of the docstring>
    """
    # ... the rest of the function ...
```
```

You should also document the deprecation in the changelog by using Reno. Explain the deprecation
and how to migrate.

> In particular situations where a deprecation or change might be a major disruptor for users, a
> *migration guide* might be needed. Once the migration guide is written and published, deprecation
> messages and documentation should link to it (use the `additional_msg: str` argument for
> `@deprecate_arg` and `@deprecate_func`).

Reviewer 2: ```suggestion
In particular situations where a deprecation or change might be a major disruptor for users, a
*migration guide* might be needed. Please write these guides in Qiskit's documentation at
https://github.com/Qiskit/documentation/tree/main/docs/api/migration-guides. Once
the migration guide is written and published, deprecation
messages and documentation should link to it (use the `additional_msg: str` argument for
`@deprecate_arg` and `@deprecate_func`).
```

> messages and documentation should link to it (use the `additional_msg: str` argument for

Reviewer: Minor style: I don't think we need to include the type hint for an argument when mentioning it in prose.

Reviewer 2: I added this a few months ago. I think it's useful for people to know what the argument is referring to. I agree it's not "necessary", but I think it's useful for only 5 extra characters.

Reviewer: I'd have thought the API documentation is a more appropriate place for this.  To me it adds noise and makes it harder to scan (especially because this is the second time it happens).
